Implement execution hooks framework

csadorf opened the pull request:
## Description

This pull request implements an execution hooks framework that enables users to execute certain functions *on start*, *on finish*, *on success*, and *on fail* of the execution of an operation.

Resolves #28, resolves #14.

## Motivation and Context

The motivation for this framework is to enable users to automatically execute certain functions with operations for example for logging and tracking purposes. For example, to log when a specific operation is executed, a user can provide the following function:

```python
@Project.operation
@Project.hooks.on_start(lambda op: logger.info(f"Executing {op}..."))
def foo(job):
    pass
```

The framework comes with a selection of pre-implemented hook systems to cover all the currently expected use cases, roughly in order of expensiveness:

1. `LogOperation` - Log basic information about the execution of operations to a log file.
2. `TrackOperations` - Keep detailed metadata records about the state of the project root directory and the operation, including directives, to a log file, optionally in conjunction with git.
3. `SnapshotProject` - Create a snapshot of the project root directory to keep track of the code used for the execution of an operation.
4. `TrackWorkspaceWithGit` - The workspace is treated as a git repository and automatically committed to before and after the execution of an operation. The can be done either on a per-job basis or workspace global.

A *hook system* is meant to describe a collection of hook functions that together achieve a specific purpose. The shipped hook systems are implemented as classes that can be installed project-wide via the respective `install_hooks` function.

## High-level API

Hooks can be installed either on a per-operation or a per-project basis. In this way users have the option to execute hook functions either with specific operations or with all operations of a project.

Furthermore, there are two ways that hooks can be installed. One way is directly in Python, for example within the `__main__` clause of the `project.py` file:
```python
# ...

if __name__ == '__main__':
    from flow.hooks import LogOperations
    LogOperations().install_hooks(Project()).main()
```
Alternatively, hooks can also be installed through the (project) configuration file:
```
project = my_project
[flow]
hooks=flow.hooks.LogOperations,flow.hooks.SnapshotProject(compress=True)
```
It is assumed that the entities provided here, are callable, i.e., are either a function or a functor. The first argument must be the instance of `FlowProject`.

## Low-level API

The `FlowProject` class has a `hooks` attribute with four lists: `on_start`, `on_finish`, `on_success`, and `on_fail`, which can be appended to. Hook functions installed in this way are executed for *all* operations.

Furthermore, hook functions can be installed for individual operations either with a decorator:
```python
@Project.operation
@Project.hook.on_start(my_hook_function)
def op(job):
    pass
```
or by passing a dictionary to the `add_operation()` function: `project.add_operation(..., hooks=dict(on_start=my_hook_function))`.

## Custom hook systems

Users can very easily implement their own hook systems and install them in similar manner.
For example:
```python
# myhooks.py
def log_op(operation):
    with open(operation.job.fn('operations.log'), 'a') as logfile:
        logfile.write(f"Executed operation '{operation}'.")

def install(project):
    project.hooks.on_success.append(log_op)
    return project
```

This could then be installed either in the project module:
```python
# ...

if __name__ == '__main__':
    import myhooks
    myhooks.install(Project()).main()
```
or via the configuration file:
```
project = myproject
[flow]
hooks=myhooks.install
```

## Types of Changes
<!-- Please select all items that apply either now or after creating the pull request: -->
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:
<!-- Please select all items that apply either now or after creating the pull request. -->
<!-- If you are unsure about any of these items, do not hesitate to ask! -->
- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

csadorf commented:
@bdice You were not assigned as reviewer, but might still be interested in this feature.

csadorf commented:
Note: This feature won't work with "exec operations", but since these are going away anyways, I think it's safe to introduce this feature regardless.

csadorf commented:
Resolves issue #28 .

vyasr commented on flow/hooks/git_workspace_tracking.py:
```diff
@@ -0,0 +1,115 @@
+    with open(os.path.join(root, '.gitignore'), mode='a+') as file:
+        file.seek(0)
+        if '.*\n' not in file.readlines():
+            file.write('.*\n')
```
I would add a logger entry or something here (or maybe in the `_get_or_init_git_repo` function`, just update the existing log entry to say that the gitignore file is being initialized to ignore hidden files) to indicate to users that a repo didn't exist, so you're making one and automatically ignoring hidden files.

vyasr replied:
It may also make sense for this to be an option (which is what we discussed the last time around with this PR). However, since it's relatively deep in the call stack and hidden from the user, just notifying the user loudly is probably good enough (and they can change it if they want), rather than trying to provide an API.

vyasr commented on flow/hooks/git_workspace_tracking.py:
```diff
@@ -0,0 +1,115 @@
+        if "nothing to commit, working tree clean" in str(error):
+            pass
+        else:
+            raise
```
This will also raise an error if there are *untracked* files in the repository (see [here](https://bitbucket.org/glotzer/signac-flow/pull-requests/35/execution-hooks/diff#comment-72666460))

csadorf replied:
Shouldn't be an issue, because we should automatically add all files (see response below).

vyasr commented on flow/hooks/git_workspace_tracking.py:
```diff
@@ -0,0 +1,115 @@
+    return {
+        'commit_id': str(repo.commit()),
+        'dirty': repo.is_dirty(),
+    }
```
Is this function meant to be a distinct hook for users? It's not used by the `TrackWorkspaceWithGit` class so I'm not quite sure what its purpose is.

vyasr commented on flow/hooks/git_workspace_tracking.py:
```diff
@@ -0,0 +1,115 @@
+        metadata = collect_metadata(operation)
+        metadata['stage'] = 'prior'
+        repo = self._get_repo(operation)
+        repo.git.add(A=True)
```
I think `u=True` is probably more appropriate, right? Users could put temporary testing scripts, or worse, temporary data files, into the root of the repository. I would prefer automatic adding to be opt-in, such that `commit_before` returns a callable and has an argument that controls whether the returned callable does a `git add -u` or a `git add -A`. I'm happy to brainstorm the best API to support such parametrizable hooks, I assume it would just be through a constructor argument of the class similar to the snapshot tracking hook class.

vyasr replied:
If we do stick with this behavior, then my concern with respect to untracked files in `_commit` is moot.

csadorf replied:
I don't think that makes sense, because then the user would need to manually switch into all workspace directories and add files.

vyasr replied:
I'm fine with tracking the workspace automatically. I'm less comfortable with tracking the project root automatically since I think that's much more susceptible to user clutter, but in the interests of reproducibility I guess it's fine to force the user to think about this and do their own cleanup rather than worrying about it ourselves.

csadorf commented on flow/hooks/util.py:
```diff
@@ -0,0 +1,21 @@
+
+def collect_metadata(operation):
+    return {
+        '_schema_version': 1,
```
I've updated the schema to include both and added comments for clarification.

csadorf commented on flow/hooks/util.py:
```diff
@@ -0,0 +1,21 @@
+        'time': time.time(),
+        'project': {
+            'path': operation.job._project.root_directory(),
+            'git': None,
```
Agreed, removed.

csadorf commented on flow/hooks/git_workspace_tracking.py:
```diff
@@ -0,0 +1,115 @@
+# Copyright (c) 2018 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from __future__ import print_function
```
Tx.

csadorf commented on flow/hooks/log_operations.py:
```diff
@@ -0,0 +1,38 @@
+            logger.setLevel(logging.DEBUG)
+
+            fh = logging.FileHandler(operation.job.fn(self._fn_logfile))
+            fh.setLevel(logging.DEBUG)
```
The handler handles *all* calls to the logging framework and will log *all* messages, including those on the *DEBUG* level.

See here: https://docs.python.org/3/howto/logging.html#handlers

vyasr replied:
But it only affects loggers that you have explicitly added the logger to via `logger.addHandler`, which in this case is `self._loggers[str(operation.job)]`, right? It won't affect other loggers, and these loggers are only used internally by this class when `log_operation` is called.

csadorf commented:
@vyasr Thank you very much for your comprehensive feedback. I'm working on addressing the issues. I was wondering whether it might make sense to split this up into introducing the framework and the individual "bundled" hook systems separately? In this way we don't stall the whole PR just because there some issues with the individual systems.

vyasr commented:
I like the idea of splitting the two PRs. However, I suspect that developing the framework is much easier if you have these example hooks to test with, so I'll leave it up to you to decide whether you want to do that split now or at the end. Since we squash PR commits, I would be fine with working on everything here until we're happy with the framework, then moving the bundled hooks to a different branch and deleting them from this one. That will lead to some scattering of comments relating to the example hooks, but I think it's a reasonable tradeoff.

mikemhenry commented on flow/project.py:
```diff
@@ -686,6 +720,33 @@ def _show_template_help_and_exit(self, template_environment, context):
+            return  # no hooks configured
+        for entry in hooks_config:
+            try:
+                m = re.match(r'^(?P<path>[\w.]+?)(\((?P<constructor>.*)\))?$', entry)
```
would it make sense to compile this regex (outside of the for loop) then use it for searches in the loop?

csadorf replied:
No, the `re` package has been optimized to automatically cache those and this loop would only be executed a few times, not hundreds or thousands of times anyways.

csadorf commented:
I merged master and I think I addressed all comments, however, this is still missing documentation. Let me know if we can move forward with this, and then I'll prepare that.

vyasr commented on flow/hooks/git_workspace_tracking.py:
```diff
@@ -0,0 +1,101 @@
+    def install_hooks(self, project):
+        project.hooks.on_start.append(self.commit_before)
+        project.hooks.on_success.append(self.commit_after)
+        project.hooks.on_fail.append(self.commit_after)
```
Why not just do `on_finish` rather than do both of these?

csadorf replied:
I think you're right, that should be equivalent.

vyasr commented on flow/hooks/log_operations.py:
```diff
@@ -0,0 +1,38 @@
+    def log_operation(self, operation, error=None):
+        if str(operation.job) in self._loggers:
+            logger = self._loggers[str(operation.job)]
+        else:
```
Why not use a try-except here?

csadorf commented:
@vyasr Could you resolve conversations that you consider resolved?

vyasr commented:
Done! Sorry, I meant to do that during the review but forgot by the end of it :)

csadorf commented:
> Done! Sorry, I meant to do that during the review but forgot by the end of it :)

No problem, thx!

vyasr commented:
@csadorf is planning to try to wrap this up by May 8th. If he cannot, at that point we've agreed that someone else should take over this project.

mikemhenry commented:
I feel like there is a lot of utility with this feature. I'm not able to take ownership and finish the effort, but I'm willing to provide review or test things out if it is at that stage.

vyasr commented:
Simon and I would like to pair program and finish this, but my data meeting (pre-defense meeting) is scheduled for August 18th so I won't be available at least until then.

vyasr commented:
@cbkerr @klywang thanks for handling the merge! Here are the next steps:
1. Install pre-commit hooks and run them locally to fix style. Then tests should start passing.
2. Go through all my outstanding inline comments. A number of them have yet to be addressed. At this stage I would only address the simpler ones (reformatting, renaming, etc).
3. Write up documentation describing the hooks (we can keep working on what you two started earlier). While doing so, try and address conceptually some of the bigger questions I asked Simon during my review. In particular, I think it will be very important to have clear distinctions between "hooks" and "hook systems", where the latter is effectively a collection of hooks along with a means for installing them. This also needs clearer delineation of operation- vs project-specific hooks.
4. Rewrite the code as needed.

cbkerr commented on flow/project.py:
```diff
@@ -3161,13 +3244,19 @@ def _register_operations(self):
 
+            # Update operation hooks
+            self._operation_hooks[name].update(Hooks.from_dict(self.hook[func]))
+
```
Fixed in our merge

cbkerr commented on flow/hooks/__init__.py:
```diff
@@ -0,0 +1,82 @@
+        "Instantiate the hook class from a dictionary."
+        hook = cls()
+        for key in mapping:
+            getattr(hook, key)[:] = list(mapping[key])
```
Improved naming in commit dc3bcdb888c14fb25730d72b0e87ae758b02fee4

klywang replied:
When we try to add an operation using the following code:
```python
from flow import FlowProject

def my_hook_func(operation):
    print(f"Start {operation.name}")

p = FlowProject()
p.add_operation('hello', cmd='echo $PATH', hook=dict(on_start=my_hook_func))
p.main()
```

and we get the following error:
```cmd
Traceback (most recent call last):
  File "/home/kelwang/Software/signac-flow/scratch.py", line 21, in <module>
    p.add_operation('hello', cmd='echo $PATH', hooks=dict(on_start=my_hook_func))
  File "/home/kelwang/Software/signac-flow/flow/project.py", line 4054, in add_operation
    self._operation_hooks[name].update(Hooks.from_dict(hooks))
  File "/home/kelwang/Software/signac-flow/flow/hooks/__init__.py", line 70, in from_dict
    getattr(hook, trigger_type)[:] = list(mapping[trigger_type])
TypeError: 'function' object is not iterable
```

When you suggested your change, did you mean that we should change `list(mapping[trigger_type])` to something else? 

Note: This works fine if we set up hooks using decorators and it worked without error.

vyasr replied:
@klywang this isn't your fault, it looks to be a bug. I've traced the problem and will try to propose a fix, but no need for you to worry about this bit for now.

csadorf commented:
@cbkerr Is this ready for another review?

bdice commented on flow/hooks/__init__.py:
```diff
@@ -0,0 +1,92 @@
+        "on_fail",
+    ]
+
+    def __init__(self, **kwargs):
```
I think we could make this more readable by avoiding `**kwargs` and defining all the hooks explicitly.

bdice commented on flow/hooks/__init__.py:
```diff
@@ -0,0 +1,92 @@
+        else:
+            super().__setattr__(name, value)
+
+    @classmethod
+    def from_dict(cls, mapping):
+        "Instantiate the hook class from a dictionary."
+        hook = cls()
+        for trigger_type in mapping:
+            getattr(hook, trigger_type)[:] = list(mapping[trigger_type])
+        return hook
```
Could this alternate constructor be replaced by an unpacking with the normal constructor?
```
Hooks.from_dict(d)
Hooks(**d)
```

bdice replied:
Resolved in #508.

bdice commented on flow/project.py:
```diff
@@ -305,6 +308,25 @@ def __iter__(self):
+        return self._operation_hooks[func]
+
+    def __getattr__(self, name):
+        class install_hook:
```
This should be 
```suggestion
        class _InstallHook:
```

bdice replied:
Resolved in #508.

bdice commented:
I read through some of this PR with @klywang. We're scheduling a time to dig deeper on this and consider the architecture of the code. I have two broad considerations:

- It seems like this PR's current design has prioritized _generality_ at the expense of readability and document-ability. I think there are some clear tradeoffs where I would prefer a less-general and more-specific way. Two examples:
  1. The supported hooks are given as a list `["on_start", ..., "on_error"]` hook attributes are assigned dynamically with `setattr`. I think it would be more readable to avoid dynamic attributes and just use `self.on_start = ...`. The supported hook types seem more like `pre` and `post` (a short list of known elements) rather than an arbitrary list that would require dynamic assignments.
  2. Some classes use `**kwargs` and define alternate constructors, rather than having a single path of entry with specifically-named parameters (and calling code that uses those names). This has a slight cost in generality but offers much-more readable code in my opinion.

- I think we may be able to simplify the "registration" side of this PR after we work through the follow-up implications of #487 and #382 and begin to reconcile class-level and instance-level data members and try to simplify FlowProject instantiation. I would like to think through those problems first, before diving into this PR further, because I think the simplifications will pay off.

csadorf commented on flow/project.py:
```diff
@@ -3161,13 +3244,19 @@ def _register_operations(self):
                 'post': post_conditions.get(func, None)}
 
+            # Update operation hooks
+            self._operation_hooks[name].update(Hooks.from_dict(self.hook[func]))
```
If I remember correctly, that's by design. I always considered hooks to be registered on a *per-instance* level, and not to be inherited. That's because hooks are something that executed for a particular runtime and unlike the DAG, not something that is *inherent* to a project's workflow.

vyasr commented:
@bdice @cbkerr @klywang most of my comments are still relevant after #482 and addressing them should make the merge easier later. I'd prioritize docs/addressing open discussions over further refactoring work on this branch while other work happens on the main branch. I definitely agree that we should aim to simplify before finalizing this feature, but I think writing docs should come first so that @cbkerr and @klywang can clearly articulate the goals of this PR first since understanding how it should work is critical to writing the best version of the code.

bdice commented on setup.py:
```diff
@@ -57,6 +57,7 @@
     entry_points={
         "console_scripts": [
             "flow = flow.__main__:main",
+            "flow-track-operations = flow.hooks.track_operations.__main__:main",
```
This main function doesn't exist.
```suggestion
```

bdice commented:
Now that #508 has been merged, we need to decide whether to implement the suggested hooks from this PR. I think the best way to do that would be to close this PR and open a new issue listing the proposed hooks and linking to this PR for reference implementations that could be adapted in separate PRs.

>The framework comes with a selection of pre-implemented hook systems to cover all the currently expected use cases, roughly in order of expensiveness:
> 1. `LogOperation` - Log basic information about the execution of operations to a log file.
> 2. `TrackOperations` - Keep detailed metadata records about the state of the project root directory and the operation, including directives, to a log file, optionally in conjunction with git.
> 3. `SnapshotProject` - Create a snapshot of the project root directory to keep track of the code used for the execution of an operation.
> 4. `TrackWorkspaceWithGit` - The workspace is treated as a git repository and automatically committed to before and after the execution of an operation. The can be done either on a per-job basis or workspace global.

bdice commented:
> @bdice This can be closed, right?

The only reason we left it open was because this PR also implements the specific hook classes that were not included in the general framework that was merged.

b-butler commented:
Closed see #637 for further discussion.
